src/Mod/Sketcher/Gui/TaskSketcherTool.cpp

```diff
@@ -0,0 +1,90 @@
+/***************************************************************************
+ *   Copyright (c) 2022 Pierre-Louis Boyer <pierrelouis.boyer@gmail.com>   *
+ *                                                                         *
+ *   This file is part of the FreeCAD CAx development system.              *
+ *                                                                         *
+ *   This library is free software; you can redistribute it and/or         *
+ *   modify it under the terms of the GNU Library General Public           *
+ *   License as published by the Free Software Foundation; either          *
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+#endif
+
+#include "TaskSketcherTool.h"
+#include <Gui/Application.h>
+#include <Gui/Document.h>
+#include <Gui/BitmapFactory.h>
+#include <Gui/ViewProvider.h>
+#include <Gui/WaitCursor.h>
+#include <Base/Tools.h>
+#include <Base/UnitsApi.h>
+
+#include <QEvent>
+
+#include "ViewProviderSketch.h"
+
+#include "SketcherToolDefaultWidget.h"
+
+using namespace SketcherGui;
+using namespace Gui::TaskView;
+
+TaskSketcherTool::TaskSketcherTool(ViewProviderSketch* sketchView)
+    : TaskBox(Gui::BitmapFactory().pixmap("document-new"), tr("Tool parameters"), true, nullptr)
+    , sketchView(sketchView)
+{
+    widget = std::make_unique<SketcherToolDefaultWidget>(this, sketchView);
+
+    this->groupLayout()->addWidget(widget.get());
+}
+
+TaskSketcherTool::~TaskSketcherTool()
+{}
+
+void TaskSketcherTool::toolChanged(const std::string& toolname)
+{
+    // TODO: Implement a factory here to get an appropriate widget from the toolname
+
+    // At this stage, we add a Default tool widget for all tools with a defined name, but this needs
+    // to change
+    if (toolname != "DSH_None") {

```

Comment:
This is even a better example of the anti-pattern.

It is ok to provide some special behaviour for "DSH_None", as it will not change over time with the introduction of each tool (plus there is not a tool that is DSH_None, but rather the absence of a tool).

However, for the rest, this if-else branching is not the way to go. We want that the special behaviour of a tool resides with the tool. The code here should be generic and achieve its function while relying on information provided by the tool.

I have considered several options. The most generic would be a full fledged widget settings factory. However, because at a time there is a single tool active (at maximum), it was probably overkill/YAGNI. 

Then, I settled for providing an NVI interface for DSH Handlers, and exposing a minimal configuration interface in view provider.  This has a relatively low footprint by comparison, and it is generic enough to move the whole widget to whereever within Sketcher Edit mode (dockable independent dialog,...). 

At the same time, this API can provide additional information, such as the specific widget (for cases where the default widget may not be appropriate), whether a widget should be visible or not, icons/title texts,...

I will post the branch.

(i) A widget settings factory
(ii)



 a kind of anti-pattern and we should always strive to avoid it.